File: client-app/src/app/core/services/reservation.service.ts
import { Injectable } from "@angular/core";
import { Reservation } from "../models/reservation";
import { Observable } from "rxjs/internal/Observable";
import { ApiService } from "../utilities/api.service";
import { ReservationPage } from "../models/reservation-page";
import { DatePipe } from "@angular/common";
import { HttpParams } from "@angular/common/http";
import { ReservationFilter } from '../models/reservation-filter';

@Injectable({
  providedIn: "root"
})
export class ReservationService {
  private url: string = "reservations";
  private reservationFilter: ReservationFilter;
  constructor(private apiService: ApiService, private datePipe: DatePipe) {}

  getAll(): Observable<Array<Reservation>> {
    return this.apiService.get(this.url);
  }

  add(model: Reservation): Observable<void> {
    return this.apiService.post(this.url + '/add', model);
  }

  edit(model: Reservation, id: number): Observable<void> {
    return this.apiService.put(this.url + "/" + id, model);
  }

  cancel(id: number): Observable<void> {
    return this.apiService.delete(this.url + "/" + id);
  }

  getReservationsByUsername(
    startIndex: number,
    pageSize: number,
    carName?: string,
    startDate?: Date,
    endDate?: Date
  ): Observable<ReservationPage> {
    this.reservationFilter = {};
    this.reservationFilter.startIndex = startIndex;
    this.reservationFilter.pageSize = pageSize;
    if (carName) {
      this.reservationFilter.carName = carName;
    }

    if (startDate) {
      this.reservationFilter.startDate = this.datePipe.transform(startDate, "yyyy-MM-dd HH:mm:ss");
    }

    if (endDate) {
      this.reservationFilter.endDate = this.datePipe.transform(endDate, "yyyy-MM-dd HH:mm:ss");
    }
    return this.apiService.getWithBody(
      this.url +
        "/user", this.reservationFilter
    );
  }
}

File: CarRental/src/main/java/com/ikubinfo/rental/service/reservation/ReservationService.java
package com.ikubinfo.rental.service.reservation;

import com.ikubinfo.rental.service.reservation.dto.ReservationModel;
import com.ikubinfo.rental.service.reservation.dto.ReservationPage;
import com.ikubinfo.rental.service.reservation.dto.ReservedDates;

import java.time.LocalDateTime;
import java.util.List;

public interface ReservationService {

    List<ReservationModel> getAll();

    ReservationPage getByUsername(int startIndex, int pageSize, String carName, String startDate,
                                  String endDate);

    List<ReservationModel> getByUsername();

    ReservationModel getById(Long id);

    List<ReservationModel> getByCar(Long carId);

    ReservationModel save(ReservationModel model);

    void edit(ReservationModel toUpdateModel, Long id);

    void cancel(Long id);

    int cancelByCarAndDate(LocalDateTime date, Long carId);

    List<ReservedDates> getReservedDatesByCar(Long carId);
}
